test_j1587_driver.py

```diff
@@ -1,66 +1,96 @@
 import multiprocessing
+import queue
+import threading
+import time
 import unittest
 
 import struct
 from hv_networks.J1708Driver import checksum
 from hv_networks.J1587Driver import J1587Driver
 from hv_networks.J1587Driver import J1708DriverFactory
 from hv_networks.J1587Driver import set_j1708_driver_factory
+from hv_networks.J1587Driver import TimeoutException
 
 
 # fake J1708Driver for testing
 class FakeJ1708Driver:
     def __init__(self):
+        self.to_rx = multiprocessing.Queue()
+        self.sent = multiprocessing.Queue()
+        self.stopped = threading.Event()
         return
 
-    to_rx = list()
-
     def add_to_rx(self, more_rx):
-        self.to_rx.extend(more_rx)
+        for thing in more_rx:
+            self.to_rx.put(thing)
 
     def read_message(self, checksum=False, timeout=0.5):
-        if len(self.to_rx) == 0:
+        if self.stopped.is_set():
+            return None
+        if self.to_rx.empty():
             return None
         else:
-            message = self.to_rx.pop()
+            message = self.to_rx.get()
             if checksum:  # NB: J1587Drive will always call with checksums=true
                 return message
             else:
                 return message[:-1]
 
-    sent = multiprocessing.Queue()
-
     def send_message(self, buf, has_check=False):
+        if self.stopped.is_set():
+            return
         msg = buf
         self.sent.put(msg)
 
     def close(self):
-        return
+        self.stopped.set()
+        self.to_rx.close()
+        self.sent.close()
+
+    def __del__(self):
+        self.close()
 
 
 class FakeJ1708Factory(J1708DriverFactory):
     def __init__(self):
+        self.a_lock = threading.Lock()
+        with self.a_lock:
+            self.memo_fake_driver = None
         super(FakeJ1708Factory, self).__init__()
-        self.memo_fake_driver = None
 
     def make(self):
-        if self.memo_fake_driver is None:
-            self.memo_fake_driver = FakeJ1708Driver()
-        return self.memo_fake_driver
+        with self.a_lock:
+            if self.memo_fake_driver is None:
+                self.memo_fake_driver = FakeJ1708Driver()
+            a = self.memo_fake_driver
+        return a
```

Comment:
Same thing. Unsure of behavior if `with` doesn't work out.

Reply:
there's no deadlocks (that I'm aware of, *gulp*) so `a` will always be (eventually) assigned.